File: 03_compiler_go/lexer/lexer.go
// go_compiler/lexer/lexer.go
package lexer

import (
	"tengri-lang/03_compiler_go/token"
	"unicode"
)
type Lexer struct {
	input        []rune
	position     int
	readPosition int
	ch           rune
}

func New(input string) *Lexer {
	l := &Lexer{input: []rune(input)}
	l.readChar()
	return l
}

func (l *Lexer) readChar() {
	if l.readPosition >= len(l.input) {
		l.ch = 0
	} else {
		l.ch = l.input[l.readPosition]
	}
	l.position = l.readPosition
	l.readPosition += 1
}

func (l *Lexer) peekChar() rune {
	if l.readPosition >= len(l.input) {
		return 0
	}
	return l.input[l.readPosition]
}

func (l *Lexer) NextToken() token.Token {
	var tok token.Token

	l.skipWhitespace() // Пропускаем шум ПЕРЕД каждым токеном

	switch l.ch {
	// Полный список Рун, Операторов и Разделителей
	case 'Π': tok = newToken(token.Runa_Func_Def, l.ch)
	case '—': tok = newToken(token.Runa_Var, l.ch)
	case 'Λ': tok = newToken(token.Runa_Const, l.ch)
	case 'Y': tok = newToken(token.Runa_If, l.ch)
	case 'Q': tok = newToken(token.Runa_True, l.ch)
	case 'I': tok = newToken(token.Runa_False, l.ch)
	case '↻': tok = newToken(token.Runa_Loop, l.ch)
	// case '→': tok = newToken(token.Runa_Return, l.ch)
	// case '⁞': tok = newToken(token.Runa_Log, l.ch)
	case '→': tok = newToken(token.ARROW, l.ch)
	case '⁞': tok = newToken(token.SEMICOLON, l.ch)
	case '□': tok = newToken(token.Runa_Type_Int, l.ch)
	case '⊡': tok = newToken(token.Runa_Type_Float, l.ch)
	case '∞': tok = newToken(token.Runa_Type_Str, l.ch)
	case '◇': tok = newToken(token.Runa_Type_Char, l.ch)
	case '≡': tok = newToken(token.Runa_Type_Collection, l.ch)
	case ':': tok = newToken(token.Op_Assign, l.ch)
	case '+': tok = newToken(token.Op_Plus, l.ch)
	case '-': tok = newToken(token.Op_Minus, l.ch)
	case '*': tok = newToken(token.Op_Multiply, l.ch)
	case '/': tok = newToken(token.Op_Divide, l.ch)
	case '>': tok = newToken(token.Op_Greater, l.ch)
	case ',': tok = newToken(token.Sep_Comma, l.ch)
	case '(': tok = newToken(token.Sep_LParen, l.ch)
	case ')': tok = newToken(token.Sep_RParen, l.ch)
	case '"': // Обработка строк
		tok.Type = token.StringLiteral
		tok.Literal = l.readString()
	// Конец файла
	case 0:
		tok.Literal = ""
		tok.Type = token.EOF
	// Если символ не опознан 
	default:
		if isLetter(l.ch) {
			tok.Literal = l.readIdentifier()
			tok.Type = token.Identifier
			return tok
		} else if unicode.IsDigit(l.ch) {
			tok.Type = token.IntLiteral
			tok.Literal = l.readNumber()
			return tok
		} else {
			tok = newToken(token.ILLEGAL, l.ch)
		}
	}

	l.readChar()
	return tok
}

// readString читает все символы до закрывающей кавычки
func (l *Lexer) readString() string {
	position := l.position + 1
	for {
		l.readChar()
		// Останавливаемся, если встретили кавычку или конец файла
		if l.ch == '"' || l.ch == 0 {
			break
		}
	}
	return string(l.input[position:l.position])
}

func (l *Lexer) skipWhitespace() {
	for {
		if unicode.IsSpace(l.ch) {
			l.readChar()
		} else if l.ch == '/' && l.peekChar() == '/' { // <-- ИСПРАВЛЕНИЕ: Правильная обработка комментариев
			for l.ch != '\n' && l.ch != 0 {
				l.readChar()
			}
		} else {
			break
		}
	}
}

// остальные вспомогательные функции 
func (l *Lexer) readIdentifier() string {
	position := l.position
	for isLetter(l.ch) {
		l.readChar()
	}
	return string(l.input[position:l.position])
}

func isLetter(ch rune) bool {
	return 'a' <= ch && ch <= 'z' || 'A' <= ch && ch <= 'Z' || ch == '_' || unicode.IsLetter(ch)
}

func (l *Lexer) readNumber() string {
	position := l.position
	for unicode.IsDigit(l.ch) {
		l.readChar()
	}
	return string(l.input[position:l.position])
}

func newToken(tokenType token.TokenType, ch rune) token.Token {
	return token.Token{Type: tokenType, Literal: string(ch)}
}

File: docs/prototype_python/lexer.py
# 02_prototype_python/lexer.py # В реальном проекте это будет в отдельном файле
from token import Token

class Lexer:
    """«Танушы» — распознающий. Финальная версия прототипа."""
    def __init__(self, source_code):
        self.code = source_code
        self.position = 0
        # Полный словарь для мгновенного распознавания одиночных символов
        self.token_map = {
            'Π': 'Runa_Func_Def', '↑': 'Runa_EntryPoint', 'Y': 'Runa_If',
            'Q': 'Runa_True', 'I': 'Runa_False', '↻': 'Runa_Loop',
            '→': 'Runa_Return', '⁞': 'Runa_Log', '—': 'Runa_Var',
            'Λ': 'Runa_Const', '□': 'Runa_Type_Int', '⊡': 'Runa_Type_Float',
            '∞': 'Runa_Type_Str', '◇': 'Runa_Type_Char', '≡': 'Runa_Type_Collection',
            '+': 'Op_Plus', '-': 'Op_Minus', '*': 'Op_Multiply', '/': 'Op_Divide',
            ':': 'Op_Assign', '@': 'Op_Access_At', '#': 'Op_Get_Size',
            '∈': 'Op_In', '(': 'Sep_LParen', ')': 'Sep_RParen',
            '[': 'Sep_LBracket', ']': 'Sep_RBracket', ',': 'Sep_Comma'
        }

    def _skip_noise(self):
        """Пропускает пробелы и комментарии."""
        while self.position < len(self.code):
            if self.code[self.position].isspace():
                self.position += 1
                continue
            if self.code[self.position:self.position+2] == '//':
                while self.position < len(self.code) and self.code[self.position] != '\n':
                    self.position += 1
                continue
            break

    def _read_identifier(self):
        """Читает идентификатор (имя)."""
        start_pos = self.position
        while self.position < len(self.code) and self.code[self.position].isalnum():
            self.position += 1
        return self.code[start_pos:self.position]

    def _read_number(self):
        """Читает целое или вещественное число."""
        start_pos = self.position
        is_float = False
        while self.position < len(self.code) and self.code[self.position].isdigit():
            self.position += 1
        if self.position < len(self.code) and self.code[self.position] == '.':
            is_float = True
            self.position += 1
            while self.position < len(self.code) and self.code[self.position].isdigit():
                self.position += 1
        
        num_str = self.code[start_pos:self.position]
        return Token('FloatLiteral', float(num_str)) if is_float else Token('IntegerLiteral', int(num_str))

    def get_next_token(self):
        """Главный метод, возвращающий следующий токен из кода."""
        self._skip_noise()

        if self.position >= len(self.code):
            return Token('EOF', None)

        current_char = self.code[self.position]

        # Проверка на составные операторы
        if current_char == '=' and self.position + 1 < len(self.code) and self.code[self.position+1] == '=':
            self.position += 2
            return Token('Op_Equal', '==')
        if current_char == '<' and self.position + 1 < len(self.code) and self.code[self.position+1] == '-':
            self.position += 2
            return Token('Op_Push', '←')

        # Проверка по словарю одиночных символов
        if current_char in self.token_map:
            token_type = self.token_map[current_char]
            token = Token(token_type, current_char)
            self.position += 1
            return token

        # Проверка на число
        if current_char.isdigit():
            return self._read_number()

        # Проверка на идентификатор
        if current_char.isalpha():
            identifier = self._read_identifier()
            return Token('Identifier', identifier)

        raise Exception(f"Ошибка: Нераспознанный символ '{current_char}'")